File: elixir/bob/bob.exs
defmodule Bob do
  defp shouting?(text) do
    text_chars = String.replace(text, ~r/[^\p{L}]/iu, "") 
    String.length(text_chars) > 0 && String.upcase(text_chars) == text_chars 
  end

  defp question?(text) do String.ends_with?(text, "?") end
  defp empty?(text) do String.length(String.trim(text)) == 0 end

  def hey(input) do
    clean_input = String.trim(input)
    cond do
      empty?(clean_input) -> "Fine. Be that way!" 
      shouting?(clean_input) && question?(clean_input) -> "Calm down, I know what I'm doing!"
      question?(clean_input) -> "Sure."
      shouting?(clean_input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end
end

File: clojure/bob/src/bob.clj
(ns bob)

(defn- question? [s]
  (clojure.string/ends-with? s "?"))

(defn- isolate-chars [s]
  (apply str (filter #(Character/isLetter %) s)))

(defn- shouted? [s]
  (let [pure-s (isolate-chars s)]
    (and
     (< 0 (count pure-s))
     (= pure-s (clojure.string/upper-case pure-s)))))

(defn- silence? [s]
  (= 0 (count (clojure.string/trim s))))

(defn- shouted-question? [s]
  (and (shouted? s) (question? s)))

(defn response-for [s]
  (cond
    (shouted-question? s) "Calm down, I know what I'm doing!"
    (question? s) "Sure."
    (silence? s) "Fine. Be that way!"
    (shouted? s) "Whoa, chill out!"
    :else "Whatever."))
